File: src/ios/objective-c/NativeOauthIds.m
#import "NativeOauthIds.h"

@implementation NativeOauthIds

- (void)login:(CDVInvokedUrlCommand*)command {
  self.loginCallbackId = command.callbackId;
  NSString *url = [command.arguments objectAtIndex:0];
  if (![url isEqualToString:@""]) {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onNotification:) name:IDNotificationManager.NOTIFICATION_CENTER_NAME object:nil];
    CieIDWKWebViewController *cieIDAuthenticator = [[CieIDWKWebViewController alloc] init];
    cieIDAuthenticator.modalPresentationStyle = UIModalPresentationFullScreen;
    cieIDAuthenticator.delegate = self;
    cieIDAuthenticator.path = url;
    [self.viewController presentViewController:cieIDAuthenticator animated:YES completion:nil];
  } else {
    [self.commandDelegate sendPluginResult:self.errorResult callbackId:command.callbackId];
  }
}

- (void)onNotification:(NSNotification*)notification {
  NSDictionary *dict = notification.userInfo;
  if (dict[@"code"] != nil && dict[@"session_state"] != nil) {
    NSString *code = dict[@"code"];
    NSString *session_state = dict[@"session_state"];
    NSString *msg = [NSString stringWithFormat:@"code=%@&session_state=%@", code, session_state];
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:msg];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:self.loginCallbackId];
  }
}

- (void)CieIDAuthenticationClosedWithSuccess {

}

- (void)CieIDAuthenticationCanceled {

}

- (void)CieIDAuthenticationClosedWithError:(NSString*)errorMessage {

}

@end

File: src/ios/NativeOauthIds.swift

@objc(NativeOauthIds) class NativeOauthIds : CDVPlugin,CieIdDelegate {

  var loginCallbackId: String = "";
    
  @objc(login:)
  func login(command: CDVInvokedUrlCommand) { 
    loginCallbackId = command.callbackId
    let url = command.arguments[0] as? String ?? ""

      if !url.isEmpty{

        NotificationCenter.default.addObserver(self, selector: #selector(self.onNotification(notification:)), name: Notification.Name(IDNotificationManager.NOTIFICATION_CENTER_NAME), object: nil)
       
        let cieIDAuthenticator = CieIDWKWebViewController()
        cieIDAuthenticator.modalPresentationStyle = .fullScreen
        cieIDAuthenticator.delegate = self
        cieIDAuthenticator.path = url;

        self.viewController?.present(
          cieIDAuthenticator,
          animated: true,
          completion: nil
        )

      }else{
          sendResultError();

      }

    }
    
    @objc func onNotification(notification: Notification){

        if let dict = notification.userInfo as Dictionary? {

            if(dict["code"] != nil && dict["session_state"] != nil){

                let code = (dict["code"]!) as! String;

                let session_state = (dict["session_state"]!) as! String;

                let msg = "code=\(code)&session_state=\(session_state)";

                let pluginResult = CDVPluginResult(
                  status: CDVCommandStatus_OK,
                  messageAs: msg
                )

                self.commandDelegate!.send(pluginResult, callbackId: loginCallbackId);

            }

        } else {
            sendResultError();
        }

    }
    
    
    @objc func  CieIDAuthenticationClosedWithSuccess() {

    }

    @objc func  CieIDAuthenticationCanceled() {
        sendResultError();
    }

    @objc func  CieIDAuthenticationClosedWithError(errorMessage: String) {
        sendResultError();
    }
    
    @objc func sendResultError(){
        let pluginResult = CDVPluginResult(
          status: CDVCommandStatus_OK,
          messageAs: "ERROR"
        )

        self.commandDelegate!.send(pluginResult, callbackId: loginCallbackId);
    }

  }

